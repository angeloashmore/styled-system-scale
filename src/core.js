const defaults = { breakpoints: [40, 52, 64].map(x => x + 'rem') }

const firstLeft = (arr, idx) => {
  if (idx >= arr.length) idx = arr.length - 1

  while (idx > -1 && (arr[idx] === null || arr[idx] === undefined)) idx--

  return arr[idx]
}

const parsePropValue = value => {
  const isNegative = value[0] === '-'
  const baseValue = isNegative ? value.slice(1) : value

  return [baseValue, isNegative]
}

const negate = v => (typeof v === 'number' ? -v : '-' + v)

export const scales = configs => {
  const cache = {}
  const parse = props => {
    cache.breakpoints =
      (props.theme && props.theme.breakpoints) || defaults.breakpoints
    const systemProps = {}

    const propKeys = Object.keys(props)

    for (let i = 0; i < propKeys.length; i++) {
      const key = propKeys[i]
      const prop = props[key]
      const config = configs[key]

      if (!config) continue

      const systemProp = config.systemProp
      const scale =
        (props.theme && props.theme[config.scale]) || config.defaultScale

      if (!Array.isArray(prop)) {
        const [s, isNeg] = parsePropValue(prop)

        if (s === undefined) continue

        if (isNeg) systemProps[systemProp] = scale[s].map(negate)
        else systemProps[systemProp] = scale[s]
      } else {
        const result = []

        for (let j = 0; j < cache.breakpoints.length + 1; j++) {
          const [s, isNeg] = parsePropValue(firstLeft(prop, j))

          if (s === undefined) continue

          if (isNeg) result[j] = negate(scale[s][j])
          else result[j] = scale[s][j]
        }

        systemProps[systemProp] = result
      }
    }

    return Object.assign(systemProps, props)
  }

  parse.configs = configs
  parse.cache = cache

  return parse
}

export const composeScales = (...parsers) => {
  const configs = {}

  for (let i = 0; i < parsers.length; i++)
    Object.assign(configs, parsers[i].configs)

  return scales(configs)
}

export const interceptScales = scaleParser => systemParser => props =>
  systemParser(scaleParser(props))
