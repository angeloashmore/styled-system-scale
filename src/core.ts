type ScalePropElement = string | number
type ScaleProp = ScalePropElement | ScalePropElement[]
type ThemeScale = Record<any, ScalePropElement[]>

type Props = Record<any, any>

interface Theme {
  breakpoints: string[]
  [key: string]: unknown
}

interface ScaleConfig {
  systemProp: string
  scale?: string
  defaultScale?: ThemeScale
}

export type ScaleConfigs = Record<any, ScaleConfig>

const DEFAULT_BREAKPOINTS = ['40rem', '52rem', '64rem']

const firstLeft = <T>(arr: T[], idx: number) => {
  if (idx >= arr.length) idx = arr.length - 1

  while (idx > -1 && (arr[idx] === null || arr[idx] === undefined)) idx--

  return arr[idx]
}

const parsePropValue = (
  value: ScalePropElement,
): [ScalePropElement, boolean] => {
  const isNumber = typeof value === 'number'
  const isNegative = isNumber
    ? (value as number) < 0
    : (value as string)[0] === '-'
  const baseValue = isNegative
    ? isNumber
      ? -(value as number)
      : (value as string).slice(1)
    : value

  return [baseValue, isNegative]
}

const negate = (v: ScalePropElement) => (typeof v === 'number' ? -v : '-' + v)

const parseProp = (prop: ScaleProp, theme: Theme, config: ScaleConfig) => {
  const scaleName = config.scale || `${config.systemProp}Scales`
  const scale =
    (theme[scaleName] as ThemeScale | undefined) || config.defaultScale
  if (!scale) return prop

  if (Array.isArray(prop)) {
    const result = []
    const count = (theme.breakpoints || DEFAULT_BREAKPOINTS).length + 1

    for (let i = 0; i < count; i++) {
      const [s, isNeg] = parsePropValue(firstLeft(prop, i))

      if (s === undefined) return

      result[i] = isNeg ? negate(scale[s][i]) : scale[s][i]
    }

    return result
  }

  const [s, isNeg] = parsePropValue(prop)

  if (s === undefined) return

  return isNeg ? scale[s].map(negate) : scale[s]
}

export interface Parser {
  (props: Props): Props
  configs: ScaleConfigs
}

export const scales = (configs: ScaleConfigs) => {
  const parse: Parser = props => {
    const propKeys = Object.keys(props)
    const systemProps: Props = {}
    const theme = props.theme || {}

    for (let i = 0; i < propKeys.length; i++) {
      if (!(propKeys[i] in configs)) continue

      const key = propKeys[i]
      const prop = props[key]
      const config = configs[key]
      const systemProp = config.systemProp

      const val = parseProp(prop, theme, config)

      systemProps[systemProp] = val
    }

    return Object.assign(systemProps, props)
  }

  parse.configs = configs

  return parse
}

export const composeScales = (...parsers: Parser[]) => {
  const configs = {}

  for (let i = 0; i < parsers.length; i++)
    Object.assign(configs, parsers[i].configs)

  return scales(configs)
}

type StyledSystemParser = (props: Props) => Props

export const interceptScales = (scaleParser: Parser) => (
  systemParser: StyledSystemParser,
) => (props: Props) => systemParser(scaleParser(props))
