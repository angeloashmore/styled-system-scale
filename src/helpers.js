import { stripUnit } from 'polished'

// Generator function that returns an increasing modular scale value on each
// `next` call.
export function* modularScaleGen(i = 0, r = 2, p = 10) {
  let j = 0
  while (true) yield Math.round(p * Math.pow(r, j++)) / p + i
}

// Generator function that returns an increasing linear scale value on each
// `next` call.
export function* linearScaleGen(i = 0, r = 1, p = 10) {
  let j = 0
  while (true) yield Math.round(p * j++ * r) / p + i
}

// Returns a linear ratio given a min, max, and number of steps to get from min
// to max.
export const linearRatio = (min, max, steps = 1) => (max - min) / steps

// Returns a scale using the provided scaling function.
export const scale = (
  count = 0,
  gen = linearScaleGen(),
  { unit = 0, min = -Infinity, max = Infinity, transform = x => x } = {},
) => {
  const arr = []

  for (let i = 0; i < count; i++)
    arr.push(
      Math.min(max, Math.max(min, transform(gen.next().value, i, count))) +
        unit,
    )

  return arr
}

export const linearScale = (
  min,
  max,
  { count, precision, unit, ...opts } = {},
) => {
  const [minV, minU] = stripUnit(min, true)
  const [maxV, maxU] = stripUnit(max, true)

  if (!unit && minU !== maxU)
    throw new Error('Units of min and max do not match.')

  count = count || maxV - minV + 1
  unit = unit || minU

  const r = linearRatio(minV, maxV, count - 1)
  const gen = linearScaleGen(minV, r, precision)

  return scale(count, gen, { unit, ...opts })
}
