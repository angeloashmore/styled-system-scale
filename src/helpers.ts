type ScaleElement = string | number | undefined
type Scale = ScaleElement[]

const _cssRegex = /^([+-]?(?:\d+|\d*\.\d+))([a-z]*|%)$/

const _stripUnit = (
  val: ScaleElement,
  unitReturn: boolean = false,
): [ScaleElement, string?] | ScaleElement => {
  if (typeof val !== 'string') return unitReturn ? [val, undefined] : val
  const matchedValue = val.match(_cssRegex)

  if (unitReturn) {
    if (matchedValue) return [parseFloat(val), matchedValue[2]]
    return [val, undefined]
  }

  if (matchedValue) return parseFloat(val)
  return val
}

// Generator function that returns an increasing modular scale value on each
// `next` call.
export function* modularScaleGen(
  i = 0,
  r = 2,
  p = 100,
): Generator<number, number, number> {
  let j = 0
  while (true) yield Math.round(p * Math.pow(r, j++)) / p + i
}

// Generator function that returns an increasing linear scale value on each
// `next` call.
export function* linearScaleGen(
  i = 0,
  r = 1,
  p = 100,
): Generator<number, number, number> {
  let j = 0
  while (true) yield Math.round(p * j++ * r) / p + i
}

// Returns a linear ratio given a min, max, and number of steps to get from min
// to max.
export const linearRatio = (min: number, max: number, steps = 1) =>
  (max - min) / steps

interface ScaleOptions {
  unit?: string
  min?: number
  max?: number
  transform?: (val: number, index: number, count: number) => number
}

// Returns a scale using the provided scaling function.
export const scale = (
  count = 0,
  gen = linearScaleGen(),
  options: ScaleOptions = {},
): Scale => {
  const {
    unit,
    min = -Infinity,
    max = Infinity,
    transform = val => val,
  } = options

  const arr = []

  for (let i = 0; i < count; i++) {
    let val: string | number = Math.min(
      max,
      Math.max(min, transform(gen.next().value, i, count)),
    )
    if (unit !== null && unit !== undefined) val = val + unit

    arr.push(val)
  }

  return arr
}

interface LinearScaleOptions extends ScaleOptions {
  count?: number
  ratio?: number
  precision?: number
}

// Returns a linear scale.
export const linearScale = (
  min: string | number,
  max: string | number,
  options: LinearScaleOptions = {},
): Scale => {
  const { precision, ...opts } = options
  let { count, ratio, unit } = options

  if (count && ratio)
    throw new Error(
      'Count and ratio were provided, but only one can be defined.',
    )

  const [minV, minU] = _stripUnit(min, true) as [number, string?]
  const [maxV, maxU] = _stripUnit(max, true) as [number, string?]

  if (!unit && minU !== maxU)
    throw new Error('Units of min and max do not match.')

  unit = unit || minU

  if (count) {
    ratio = linearRatio(minV, maxV, count - 1)
  } else if (ratio) {
    count = (maxV - minV) / ratio + 1
  } else {
    count = maxV - minV + 1
    ratio = 1
  }

  const gen = linearScaleGen(minV, ratio, precision)

  return scale(count, gen, { unit, min: minV, max: maxV, ...opts })
}

// Negates all values of a scale. Maintains units.
export const negateScale = (scale: Scale) =>
  scale.map(x => {
    if (typeof x === undefined) return x

    const [v, unit] = _stripUnit(x, true) as [string | number, string?]

    return unit === undefined ? -v : -v + unit
  })

// Adds two scales together. Maintains the unit of the left array. Unit of the
// right array is ignored.
export const addScales = (a: Scale, b: Scale): Scale => {
  const [, unit] = _stripUnit(a[0], true) as [unknown, string?]
  const hasUnit = unit !== undefined
  const result = []

  for (let i = 0; i < a.length; i++) {
    const ai = _stripUnit(a[i] || 0) as string | number
    const bi = _stripUnit(b[i] || 0) as string | number

    result[i] = hasUnit
      ? (ai as string) + (bi as string) + unit
      : (ai as number) + (bi as number)
  }

  return result
}

// Subtracts two scales. Maintains the unit of the left array. Unit of the
// right array is ignored.
export const subtractScales = (a: Scale, b: Scale) =>
  addScales(a, negateScale(b))

// Merge scale values from the right scale into the left scale. An undefined
// value in the right scale falls back to the left scale value.
export const mergeScalesLeft = (a: Scale, b: Scale): Scale => {
  const result = []

  for (let i = 0; i < Math.max(a.length, b.length); i++)
    result[i] = b[i] === undefined ? a[i] : b[i]

  return result
}

// Merge scale values from the left scale into the right scale. An undefined
// value in the left scale falls back to the right scale value.
export const mergeScalesRight = (a: Scale, b: Scale) => mergeScalesLeft(b, a)
