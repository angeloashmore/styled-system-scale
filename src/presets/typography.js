import { scales } from '../index'

const configs = {
  fontSizeScale: {
    systemProp: 'fontSize',
    scale: 'fontSizeScales',
  },
}

export const typographyScales = scales(configs)
