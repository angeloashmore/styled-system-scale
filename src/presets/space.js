import { scales } from '../core'

const configs = {}

configs.margin = {
  marginScale: {
    systemProp: 'margin',
    scale: 'spaceScales',
  },
  marginTopScale: {
    systemProp: 'marginTop',
    scale: 'spaceScales',
  },
  marginRightScale: {
    systemProp: 'marginRight',
    scale: 'spaceScales',
  },
  marginBottomScale: {
    systemProp: 'marginBottom',
    scale: 'spaceScales',
  },
  marginLeftScale: {
    systemProp: 'marginLeft',
    scale: 'spaceScales',
  },
  marginXScale: {
    properties: 'marginX',
    scale: 'spaceScales',
  },
  marginYScale: {
    systemProp: 'marginY',
    scale: 'spaceScales',
  },
}
configs.margin.mScale = configs.margin.marginScale
configs.margin.mtScale = configs.margin.marginTopScale
configs.margin.mrScale = configs.margin.marginRightScale
configs.margin.mbScale = configs.margin.marginBottomScale
configs.margin.mlScale = configs.margin.marginLeftScale
configs.margin.mxScale = configs.margin.marginXScale
configs.margin.myScale = configs.margin.marginYScale

configs.padding = {
  paddingScale: {
    systemProp: 'padding',
    scale: 'spaceScales',
  },
  paddingTopScale: {
    systemProp: 'paddingTop',
    scale: 'spaceScales',
  },
  paddingRightScale: {
    systemProp: 'paddingRight',
    scale: 'spaceScales',
  },
  paddingBottomScale: {
    systemProp: 'paddingBottom',
    scale: 'spaceScales',
  },
  paddingLeftScale: {
    systemProp: 'paddingLeft',
    scale: 'spaceScales',
  },
  paddingXScale: {
    systemProp: 'paddingX',
    scale: 'spaceScales',
  },
  paddingYScale: {
    systemProp: 'paddingY',
    scale: 'spaceScales',
  },
}
configs.padding.pScale = configs.padding.paddingScale
configs.padding.ptScale = configs.padding.paddingTopScale
configs.padding.prScale = configs.padding.paddingRightScale
configs.padding.pbScale = configs.padding.paddingBottomScale
configs.padding.plScale = configs.padding.paddingLeftScale
configs.padding.pxScale = configs.padding.paddingXScale
configs.padding.pyScale = configs.padding.paddingYScale

export const marginScales = scales(configs.margin)
export const paddingScales = scales(configs.padding)
export const spaceScales = scales({ ...configs.margin, ...configs.padding })
