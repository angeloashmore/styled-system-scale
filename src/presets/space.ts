import { ThemeValue, ResponsiveValue } from 'styled-system'

import { ScalesTheme, RequiredScalesTheme } from '..'
import { scales, ScaleConfigs } from '../core'

/**
 * A convenience scale group containing props related to margin and padding.
 */
export interface SpaceScaleProps<
  ThemeType extends ScalesTheme = RequiredScalesTheme,
  TVal = ThemeValue<'spaceScales', ThemeType>
> {
  /** Margin on top, left, bottom and right */
  mScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on top, left, bottom and right */
  marginScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on top */
  mtScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on top */
  marginTopScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on right */
  mrScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on right */
  marginRightScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on bottom */
  mbScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on bottom */
  marginBottomScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on left */
  mlScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on left */
  marginLeftScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on left and right */
  mxScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on left and right */
  marginXScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on top and bottom */
  myScale?: ResponsiveValue<TVal, ThemeType>
  /** Margin on top and bottom */
  marginYScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on top, left, bottom and right */
  pScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on top, left, bottom and right */
  paddingScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on top */
  ptScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on top */
  paddingTopScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on right */
  prScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on right */
  paddingRightScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on bottom */
  pbScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on bottom */
  paddingBottomScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on left */
  plScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on left */
  paddingLeftScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on left and right */
  pxScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on left and right */
  paddingXScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on top and bottom */
  pyScale?: ResponsiveValue<TVal, ThemeType>
  /** Padding on top and bottom */
  paddingYScale?: ResponsiveValue<TVal, ThemeType>
}

const configs: Record<any, ScaleConfigs> = {}

configs.margin = {
  marginScale: {
    systemProp: 'm',
    scale: 'spaceScales',
  },
  marginTopScale: {
    systemProp: 'mt',
    scale: 'spaceScales',
  },
  marginRightScale: {
    systemProp: 'mr',
    scale: 'spaceScales',
  },
  marginBottomScale: {
    systemProp: 'mb',
    scale: 'spaceScales',
  },
  marginLeftScale: {
    systemProp: 'ml',
    scale: 'spaceScales',
  },
  marginXScale: {
    systemProp: 'mx',
    scale: 'spaceScales',
  },
  marginYScale: {
    systemProp: 'my',
    scale: 'spaceScales',
  },
}
configs.margin.mScale = configs.margin.marginScale
configs.margin.mtScale = configs.margin.marginTopScale
configs.margin.mrScale = configs.margin.marginRightScale
configs.margin.mbScale = configs.margin.marginBottomScale
configs.margin.mlScale = configs.margin.marginLeftScale
configs.margin.mxScale = configs.margin.marginXScale
configs.margin.myScale = configs.margin.marginYScale

configs.padding = {
  paddingScale: {
    systemProp: 'p',
    scale: 'spaceScales',
  },
  paddingTopScale: {
    systemProp: 'pt',
    scale: 'spaceScales',
  },
  paddingRightScale: {
    systemProp: 'pr',
    scale: 'spaceScales',
  },
  paddingBottomScale: {
    systemProp: 'pb',
    scale: 'spaceScales',
  },
  paddingLeftScale: {
    systemProp: 'pl',
    scale: 'spaceScales',
  },
  paddingXScale: {
    systemProp: 'px',
    scale: 'spaceScales',
  },
  paddingYScale: {
    systemProp: 'py',
    scale: 'spaceScales',
  },
}
configs.padding.pScale = configs.padding.paddingScale
configs.padding.ptScale = configs.padding.paddingTopScale
configs.padding.prScale = configs.padding.paddingRightScale
configs.padding.pbScale = configs.padding.paddingBottomScale
configs.padding.plScale = configs.padding.paddingLeftScale
configs.padding.pxScale = configs.padding.paddingXScale
configs.padding.pyScale = configs.padding.paddingYScale

export const marginScales = scales(configs.margin)
export const paddingScales = scales(configs.padding)
export const spaceScales = scales({ ...configs.margin, ...configs.padding })
