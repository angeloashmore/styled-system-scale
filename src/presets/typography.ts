import { ThemeValue, ResponsiveValue } from 'styled-system'

import { RequiredScalesTheme, ScalesTheme } from '..'
import { scales } from '../core'

/**
 * A convenience scale group containing props related to typography.
 */
export interface TypographyScaleProps<
  ThemeType extends ScalesTheme = RequiredScalesTheme,
  TVal = ThemeValue<'fontSizeScales', ThemeType>
> {
  /**
   * The font-size CSS property sets the size of the font. This property is
   * also used to compute the size of em, ex, and other relative <length>
   * units.
   *
   * [MDN reference](http://developer.mozilla.org/en-US/docs/Web/CSS/font-size)
   */
  fontSizeScale?: ResponsiveValue<TVal, ThemeType>
}

const configs = {
  fontSizeScale: {
    systemProp: 'fontSize',
    scale: 'fontSizeScales',
  },
}

export const typographyScales = scales(configs)
