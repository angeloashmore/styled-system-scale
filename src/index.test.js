import React from 'react'
import renderer from 'react-test-renderer'
import serializer from 'jest-emotion'
import styled from '@emotion/styled'
import { ThemeProvider } from 'emotion-theming'
import { compose, typography, space } from 'styled-system'

import { interceptScales, composeScales } from './core'
import { typographyScales } from './presets/typography'
import { spaceScales } from './presets/space'

import { linearScale, scale, linearScaleGen } from './helpers'

expect.addSnapshotSerializer(serializer)

const theme = {
  breakpoints: linearScale('40rem', '64rem', { count: 3 }),
  space: linearScale(0, 10, { count: 41 }),
  spaceScales: {
    small: linearScale(1, 4),
    base: linearScale(4, 7),
    large: linearScale(8, 11),
  },
  fontSizes: linearScale('1rem', '6rem', { count: 24, precision: 100 }),
  fontSizeScales: {
    small: linearScale(0, 3),
    base: linearScale(1, 4),
    large: linearScale(8, 11),
  },
}

const scales = composeScales(typographyScales, spaceScales)

const Comp = styled('div')(
  interceptScales(scales)(
    compose(
      typography,
      space,
    ),
  ),
)

test('theme', () => {
  expect(theme).toMatchSnapshot()
})

test.skip('provides responsive values to style prop', () => {
  const tree = renderer.create(
    <ThemeProvider theme={theme}>
      <Comp
        fontSizeScale={['medium', null, 'large']}
        pb={2}
        ptScale="medium"
        pxScale="large"
      />
    </ThemeProvider>,
  )

  expect(tree).toMatchSnapshot()
})
