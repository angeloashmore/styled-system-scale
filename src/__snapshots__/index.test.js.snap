// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provides responsive values to style prop 1`] = `
.emotion-0 {
  font-size: 0;
  padding-top: 0;
  padding-left: 2.5px;
  padding-right: 2.5px;
  padding-bottom: 0.5px;
}

@media screen and (min-width:1rem) {
  .emotion-0 {
    font-size: 1px;
    padding-top: 0.25px;
    padding-left: 2.75px;
    padding-right: 2.75px;
  }
}

@media screen and (min-width:2rem) {
  .emotion-0 {
    font-size: 12px;
    padding-top: 0.5px;
    padding-left: 3px;
    padding-right: 3px;
  }
}

@media screen and (min-width:3rem) {
  .emotion-0 {
    font-size: 13px;
    padding-top: 0.75px;
    padding-left: 3.25px;
    padding-right: 3.25px;
  }
}

<div
  className="emotion-0"
/>
`;

exports[`theme 1`] = `
Object {
  "breakpoints": Array [
    "40rem",
    "52rem",
    "64rem",
  ],
  "fontSizeScales": Object {
    "base": Array [
      1,
      2,
      3,
      4,
    ],
    "large": Array [
      8,
      9,
      10,
      11,
    ],
    "small": Array [
      0,
      1,
      2,
      3,
    ],
  },
  "fontSizes": Array [
    "0.75rem",
    "0.95rem",
    "1.15rem",
    "1.35rem",
    "1.55rem",
    "1.85rem",
    "2.05rem",
    "2.25rem",
    "2.45rem",
    "2.65rem",
    "2.85rem",
    "3.05rem",
    "3.25rem",
    "3.45rem",
    "3.65rem",
    "3.95rem",
    "4.15rem",
    "4.35rem",
    "4.55rem",
    "4.75rem",
    "4.95rem",
    "5.15rem",
    "5.35rem",
    "5.55rem",
    "5.75rem",
    "6.05rem",
  ],
  "space": Array [
    0,
    0.3,
    0.5,
    0.8,
    1,
    1.3,
    1.5,
    1.8,
    2,
    2.3,
    2.5,
    2.8,
    3,
    3.3,
    3.5,
    3.8,
    4,
    4.3,
    4.5,
    4.8,
    5,
    5.3,
    5.5,
    5.8,
    6,
    6.3,
    6.5,
    6.8,
    7,
    7.3,
    7.5,
    7.8,
    8,
    8.3,
    8.5,
    8.8,
    9,
    9.3,
    9.5,
    9.8,
    10,
  ],
  "spaceScales": Object {
    "base": Array [
      4,
      5,
      6,
      7,
    ],
    "large": Array [
      8,
      9,
      10,
      11,
    ],
    "small": Array [
      1,
      2,
      3,
      4,
    ],
  },
}
`;
